src/k8s/pkg/utils/mapstructure.go

```diff
@@ -42,3 +42,45 @@ func StringToFieldsSliceHookFunc(r rune) mapstructure.DecodeHookFunc {
 		return strings.FieldsFunc(raw, func(this rune) bool { return this == r || unicode.IsSpace(this) }), nil
 	}
 }
+
+// YAMLToStringMapHookFunc returns a mapstructure.DecodeHookFunc that converts string to map[string]string by parsing YAML.
+func YAMLToStringMapHookFunc(f reflect.Kind, t reflect.Kind, data interface{}) (interface{}, error) {
+	if f != reflect.String || t != reflect.Map {
+		return data, nil
+	}
+
+	if data.(string) == "" {
+		return map[string]string{}, nil
+	}
+
+	var result map[string]string
+	if err := yaml.Unmarshal([]byte(data.(string)), &result); err != nil {
+		return data, nil
+	}
+
+	return result, nil
+}
+
+// StringToStringMapHookFunc is like StringToFieldsSliceHookFunc(). It splits map entries on ',' and then key-value pairs on '='.
+func StringToStringMapHookFunc(f reflect.Kind, t reflect.Kind, data interface{}) (interface{}, error) {
+	if f != reflect.String || t != reflect.Map {
+		return data, nil
+	}
+
+	raw := data.(string)
+	if raw == "" {
+		return map[string]string{}, nil
+	}
+
+	fields := strings.FieldsFunc(raw, func(this rune) bool { return this == ',' || unicode.IsSpace(this) })
+	result := make(map[string]string, len(fields))
+	for _, kv := range strings.FieldsFunc(raw, func(this rune) bool { return this == ',' || unicode.IsSpace(this) }) {
+		parts := strings.SplitN(kv, "=", 2)
+		if len(parts) < 2 {
+			return data, nil
```

Comment:
I'm a bit confused about this function. Why do you return the data here? 
For `key1 value1, key2=value2` this will return `key1 value1, key2=value2` without any error.
Also, we seem to allow spaces (line 75) but this doesn't work with this check (e.g. ` key1 = value1 , key2 = value2 ` returns ` key1 = value1 , key2 = value2 `).
Please add some unittests to cover this function and the one above.
You can take my tests from here: https://go.dev/play/p/49a0buG-qEj

Reply:
when parsing arbitrary `k1=v1, k2=v2` string inputs here, `kv` will be `k1=v1` and `k2=v2`.

then, we split into at most 2 parts on `=`. The result of that is:

```
strings.SplitN("k1", "=", 2) ==> []string{"k1"}
strings.SplitN("k1=v1", "=", 2) ==> []string{"k1", "v1"}
strings.SplitN("k1=v1=v2", "=", 2) ==> []string{"k1", "v1=v2"} 
```

the check for `len(parts) < 2` handles the first case (i.e. there is no `=` in the string. in that case, the decode hook has failed, therefore we `return data, nil` so that we fallthrough to the next hook (if any).

Reply:
I know it's not pretty but can we please include this example in a comment right there?

Reply:
the test cases should account for this https://github.com/canonical/k8s-snap/pull/459/files#diff-9f869847ae70a71fcb8c1ca0ad5adeb8caec20026961223a709bc6361a179ba4R82